### PR TITLE
[#97604240] Fix shadow texture uniform warning

### PR DESCRIPTION
**Pull request by Kuranes**
Fix useless shadow texture binding during cast
Removes useless uniforms for none/pcf
Add shadow offset rotation

### CODE REVIEW

**Comment by Kuranes** [comment]
(Not Ready For Merge)

File: sources/osg/State.js [hunk 1]
```diff
@@ -515,8 +515,10 @@ define( [
             // if the the stack has changed but the last applied attribute is the same
             // then we dont need to apply it again
             if ( attributeStack.lastApplied !== attribute ) {
-                gl.activeTexture( gl.TEXTURE0 + textureUnit );
-                attribute.apply( this, textureUnit );
+                if ( !attribute._textureNull ) {
```
Comment (cedricpinson) [inline sources/osg/State.js]: Adds a methods to Texture attribute like isTextureNull then use this method for the test

Comment (cedricpinson) [inline sources/osg/State.js]: Also you can split this commit to another PR because it's easy to merge it but maybe we will spend more time on shadow review

File: sources/osgShader/Compiler.js [hunk 2]
```diff
@@ -317,10 +317,12 @@ define( [
...
             if ( exist ) {
                 // see comment in Variable function
-                if ( exist.getType() !== type ) {
+                if ( exist.getType() === type ) {
+                    return exist;
+                }
+                if ( exist.getType() === 'sampler2D' && type !== 'sampler2D' ) {
                     Notify.error( 'Same uniform, but different type' );
                 }
-                return exist;
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: it means that if it's not the same type we dont return exist anymore, what is the consequence of changing that ?

File: sources/osgShader/Compiler.js [hunk 3]
```diff
@@ -627,21 +629,14 @@ define( [
...
-                    // return now to prevent creation of useless FragTexCoord
-                    //( shadow creates its own texcoord)
-                    return;
-                }
-
-
+                textureSampler = this.getOrCreateSampler( 'sampler2D', samplerName );
+            } else if ( texture.className() === 'TextureCubeMap' ) {
+                textureSampler = this.getOrCreateSampler( 'samplerCube', samplerName );
+            } else if ( texture.className() === 'ShadowTexture' ) {
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: I guess it's done somewhere else but document it

File: sources/osgShader/Compiler.js [hunk 4]
```diff
@@ -748,13 +743,22 @@ define( [
 
                 shadowTexture = shadowTextures[ k ];
                 if ( shadowTexture ) {
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: Can we create a function ?

File: sources/osgShader/Compiler.js [hunk 5]
```diff
@@ -748,13 +743,22 @@ define( [
...
-                    inputs.shadowTexture = tex;
+                    var shadowTexSamplerName = 'ShadowTexture' + k;
+
 
                     // per texture uniforms
                     var shadowTextureUniforms = this.getOrCreateTextureStateAttributeUniforms( shadowTexture, 'shadowTexture', k );
+
+                    // TODO: UGLY REMOVAL
+                    // create a specific texture getuniform for compiler
+                    shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ] = undefined;
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: how could we improve this ?

File: sources/osgShadow/ShadowMap.js [hunk 6]
```diff
@@ -420,45 +420,45 @@ define( [
 
             this._shadowReceiveAttribute.setLightNumber( lightNumber );
 
-            this._receivingStateset.setAttributeAndModes( this._shadowReceiveAttribute, StateAttribute.ON | StateAttribute.OVERRIDE );
 
 
+            var fullOverride = StateAttribute.OVERRIDE | StateAttribute.ON;
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMap.js]: can you explain ?

File: sources/osgShadow/ShadowMap.js [hunk 7]
```diff
@@ -420,45 +420,45 @@ define( [
...
+            var shouldGetMaxTextureUnits = 32;
             //
-            // TODO:: Should handle the alpha_mask that uses a texture
-            // case somehow...
+            // TODO:: Should handle the alpha_mask that uses a texture case somehow...
             // deduce from shader compil ?
-            //for ( var k = 0; k < shouldGetMaxTextureUnits; k++ ) {
-            //    this._casterStateSet.setTextureAttributeAndModes( k, blankTexture, StateAttribute.OFF | StateAttribute.OVERRIDE | StateAttribute.PROTECTED );
-            //}
+            for ( var k = 0; k < shouldGetMaxTextureUnits; k++ ) {
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMap.js]: maybe split this part into a function that says what it does, just to understand better. This part is to 'optimize' avoid texture binding right ? so maybe this could be done in a function with an appropriate name

File: sources/osgShadow/ShadowTexture.js [hunk 8]
```diff
@@ -77,8 +77,9 @@ define( [
 
             }.bind( this ) );
 
-
+            // tells shader where to find it
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowTexture.js]: I dont understand the extra logic, could you explains why it's not like Texture attribute ?

File: sources/osgShadow/shaders/tapPCF.glsl [hunk 9]
```diff
@@ -2,12 +2,44 @@
 
 #pragma include "shadowLinearSoft.glsl"
 
+float hash1D(float n) { return fract(sin(n) * 1e4); }
+float hash2D(vec2 p) { return fract(1e4 * sin(17.0 * p.x + p.y * 0.1) * (0.1 + abs(sin(p.y * 13.0 + p.x)))); }
+
+
+float noise2D(vec2 x) {
```
Comment (cedricpinson) [inline sources/osgShadow/shaders/tapPCF.glsl]: is noise2D general ? if yes it should go in a different glsl file, else it should be prefixed by something like shadow something

File: sources/osg/State.js (hunk 1 above)
Comment (Kuranes) [inline sources/osg/State.js]: ok

File: sources/osgShader/Compiler.js (hunk 5 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: ShadowTexture is the first Texture inherited class with uniform attached.
So when we get those for the compiler, we get uniform we don't want (the int 
texture location for shader texture pointer to bind)

That Int uniform conflicts with the sample2D uniform in our compiler.

Either we make compiler handle that case (not easy) or we make sure we don't do those conflicts (meaning a special getOrCreateUniform, with a parameter like "getShaderTextureLocation" flag at false when called from compiler, true when called from other code)

File: sources/osgShader/Compiler.js (hunk 4 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: k

File: sources/osgShadow/ShadowMap.js (hunk 6 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: Idea is to make sure the null texture is "binded" to all shadow casting scene, so we override all existing texture bind in the scene, preventing any texture bind.

When user implements alpha mask casting, they use PROTECTED to prevent OVERRIDE to remove their alpha mask texture needed.

File: sources/osgShadow/ShadowMap.js (hunk 7 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: k

File: sources/osgShadow/ShadowMap.js (hunk 6 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: I'll add a comment

File: sources/osgShadow/ShadowTexture.js (hunk 8 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowTexture.js]: it's the same apart from the texture name change to reflect the class name ?

File: sources/osgShadow/shaders/tapPCF.glsl (hunk 9 above)
Comment (Kuranes) [inline sources/osgShadow/shaders/tapPCF.glsl]: ok, will move it

File: sources/osg/State.js (hunk 1 above)
Comment (Kuranes) [inline sources/osg/State.js]: https://github.com/cedricpinson/osgjs/pull/412

File: sources/osgShader/Compiler.js (hunk 2 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: error is breaking and the user will get errors on compilation if the node is really used.
(in our case the node isn't used in compilation...)

File: sources/osgShadow/ShadowMap.js (hunk 7 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: splitter in the pr https://github.com/cedricpinson/osgjs/pull/412

File: sources/osgShadow/ShadowTexture.js (hunk 8 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowTexture.js]: Note: that's actually the piece of code that cause problem. 
(as the compiler will make a uniform node out if it which will conflict with the sample2D uniform)

File: sources/osgShader/Compiler.js (hunk 4 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: done

File: sources/osgShadow/shaders/tapPCF.glsl (hunk 9 above)
Comment (Kuranes) [inline sources/osgShadow/shaders/tapPCF.glsl]: done

File: sources/osgShader/Compiler.js (hunk 3 above)
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: could you add a few comment about it

File: sources/osgShader/ShaderProcessor.js [hunk 10]
```diff
@@ -116,7 +116,7 @@ define( [
                     // if not defined we do not include
                     var found = false;
                     var defines = inputsDefines.map( function ( defineString ) {
-                        return _self._defineR.test( defineString ) && defineString.split( ' ' )[ 1 ];
+                        return _self._defineR.test( defineString ) && defineString.replace( /\s+/g, ' ' ).split( ' ' )[ 1 ];
```
Comment (cedricpinson) [inline sources/osgShader/ShaderProcessor.js]: why it changed ? adds a comment plz

File: sources/osgShader/Compiler.js [hunk 11]
```diff
@@ -742,25 +737,11 @@ define( [
...
-                    var shadowTextureUniforms = this.getOrCreateTextureStateAttributeUniforms( shadowTexture, 'shadowTexture', k );
-                    inputs = MACROUTILS.objectMix( inputs, shadowTextureUniforms );
-
-                    var shadowVarying = {
-                        vertexWorld: vertexWorld,
-                        lightEyeDir: inputs.lightEyeDir,
-                        lightNDL: inputs.lightNDL
-                    };
-                    inputs = MACROUTILS.objectMix( inputs, shadowVarying );
+                    this.shadowTextureCreateUniformsInputAndVarying( shadowTexture, inputs, vertexWorld, k );
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: use the same scheme of naming method something like
createShadowTextureAndVarying
createShadowTextureInput
maybe better

File: sources/osgShader/node/noise.glsl [hunk 12]
```diff
@@ -0,0 +1,25 @@
+
+float hash1D(float n) { return fract(sin(n) * 1e4); }
```
Comment (cedricpinson) [inline sources/osgShader/node/noise.glsl]: is there some links to understand this ?

File: sources/osgShader/Compiler.js (hunk 3 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: it's part of the thing that need refactor/thinking:
Sampler2D and Int shader Texture location uniform are actually both sharing the same name thus breaking compiler uniform duplicate detection of same uniform with different types

File: sources/osgShadow/ShadowTexture.js (hunk 8 above)
Comment (cedricpinson) [inline sources/osgShadow/ShadowTexture.js]: Ok we should probably talk about this. I will spend more time to understand this part.

File: sources/osgShadow/shaders/tapPCF.glsl [hunk 13]
```diff
@@ -1,5 +1,7 @@
 
 
+#pragma include "noise.glsl" "_ROTATEOFFSET"
```
Comment (cedricpinson) [inline sources/osgShadow/shaders/tapPCF.glsl]: Could you improve the coding style with the glsl part and the meaning of _ in defines plz

Comment (Kuranes) [inline sources/osgShadow/shaders/tapPCF.glsl]: done

File: sources/osgShader/ShaderProcessor.js (hunk 10 above)
Comment (Kuranes) [inline sources/osgShader/ShaderProcessor.js]: done

File: sources/osgShader/node/noise.glsl (hunk 12 above)
Comment (Kuranes) [inline sources/osgShader/node/noise.glsl]: done updated with many links and source

File: sources/osgShader/Compiler.js (hunk 11 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: done

**Comment by cedricpinson** [comment]
needs a rebase

File: sources/osgShader/Compiler.js (hunk 3 above)
Comment (stephomi) [inline sources/osgShader/Compiler.js]: Instead of ``=== undefined``, you can just write ``!textureSampler`` (it's not like we are testing a string or a number)
Also, looks like it's broken. We can't declare a samplerCube anymore.... since it will fallback in the first branch.
And, why do we always test for ``getVariable`` when we only do a ``getOrCreate`` just after?

In sketchfab we don't use ``declareTexture`` and ``declareTextures`` because they declare too much nodes even when it's not needed.
Ping me back so that we can discuss about re-use of sketchab compiler code about texture management.

Comment (Kuranes) [inline sources/osgShader/Compiler.js]: - *!textureSampler* :+1:  
- *broken* :seems a if disapeared here, will fix 
- *why do we always test for getVariable when we only do a getOrCreate just after?* Cannot Agree More ! Especially if we already does that in other production code... Those "Declare" are not flexible at all.

File: sources/osgShader/Compiler.js [hunk 14]
```diff
@@ -775,6 +783,34 @@ define( [
...
+            // UGLY REMOVAL
+            // TODO: create a specific texture getuniform for compiler
+            var id = shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ].getID();
+            shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ] = undefined;
+            this._variables[ shadowTexSamplerName ] = undefined;
+            this._activeNodeList[ id ] = undefined;
+            delete this._activeNodeList[ id ];
+            // end UGLY REMOVAL
+
+            inputs = MACROUTILS.objectMix( inputs, shadowTextureUniforms );
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: dont override arguments variable, use a different name. not sure it works when strict is used. To avoid any doubt use a different variable name

File: sources/osgShader/ShaderProcessor.js [hunk 15]
```diff
@@ -116,7 +116,8 @@ define( [
                     // if not defined we do not include
                     var found = false;
                     var defines = inputsDefines.map( function ( defineString ) {
-                        return _self._defineR.test( defineString ) && defineString.split( ' ' )[ 1 ];
+                        // find '#define', remove duplicate whitespace, split on space and return the define Text
+                        return _self._defineR.test( defineString ) && defineString.replace( /\s+/g, ' ' ).split( ' ' )[ 1 ];
```
Comment (cedricpinson) [inline sources/osgShader/ShaderProcessor.js]: _self is a bad name variable but I know it was here before

File: sources/osgShader/node/shadows.js [hunk 16]
```diff
@@ -13,7 +13,7 @@ define( [
 
     ShadowReceive.prototype = MACROUTILS.objectInherit( Node.prototype, {
         type: 'ShadowReceiveNode',
-        validInputs: [ 'lighted', 'shadowTexture', 'shadowTextureMapSize', 'shadowTextureProjectionMatrix', 'shadowTextureViewMatrix', 'shadowTextureDepthRange', 'lightNDL', 'vertexWorld', 'shadowbias' /* 'shadowexponent0', 'shadowexponent1', 'shadowepsilonVSM' */ ],
+        validInputs: [ 'lighted', 'shadowTexture', 'shadowTextureMapSize', 'shadowTextureProjectionMatrix', 'shadowTextureViewMatrix', 'shadowTextureDepthRange', 'lightNDL', 'vertexWorld', 'shadowbias', /* 'shadowexponent0', 'shadowexponent1', 'shadowepsilonVSM' */ ],
```
Comment (cedricpinson) [inline sources/osgShader/node/shadows.js]: if not used shadowexponent0 the put the comment outside the array.

Comment (Kuranes) [inline sources/osgShader/node/shadows.js]: it's used but not mandatory (EVSM/VSM/ESM only), that's the idea of the comment.

File: sources/osgShader/Compiler.js (hunk 14 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: k

**Comment by cedricpinson** [comment]
this is not fixed https://github.com/cedricpinson/osgjs/pull/411/files#r39389990

File: sources/osgShadow/shaders/hash.glsl [hunk 17]
```diff
@@ -1,24 +1,52 @@
...
-float hashNormalized11( const in float n )
+// Dave Hoskins: hash without sin
+//https://www.shadertoy.com/view/4djSRW
+#define MOD2 vec2(443.8975,397.2973)
+//----------------------------------------------------------------------------------------
+//  1 out, 1 in...
+//note: normalized uniform random, [0;1[
+float hash11(const in float p)
+{
+  vec2 p2 = fract(vec2(p) * MOD2);
```
Comment (cedricpinson) [inline sources/osgShadow/shaders/hash.glsl]: this file use 2 spaces for indentation but this one https://github.com/cedricpinson/osgjs/pull/411/files#diff-f88a75996357a297e59617395fc78b26R8 use 4 spaces. Looks like we have the same problem on osgjs, could you use 4 spaces for all glsl files plz ?

Comment (Kuranes) [inline sources/osgShadow/shaders/hash.glsl]: arf, editorconfig was 2 spaces... updated with 4 spaces.

File: sources/osgShadow/ShadowMap.js [hunk 18]
```diff
@@ -236,7 +236,7 @@ define( [
 
         // Overridable Visitor so that user can override the visitor to enable disable
         // in its own shadowmap implementation
-        this._removeNodesNeverCastingVisitor = new RemoveNodesNeverCastingVisitor( this._castsShadowTraversalMask );
+        this._removeNodesNeverCastingVisitor = settings.userCallbackNodeCaster || new RemoveNodesNeverCastingVisitor( this._castsShadowTraversalMask );
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMap.js]: It's better to use setter, it's more the way used in osg/js to change content. do you use settings.userCallbackNodeCaster somewhere in the code ?

Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: no other place than that.

We talked about ShadowSettings and agreed it was a to ok be used without setters as it's a sort of dictionnary

File: sources/osgShadow/ShadowTexture.js [hunk 19]
```diff
@@ -77,17 +77,25 @@ define( [
...
+            // tells Shader Program where to find it
             var name = 'ShadowTexture' + unit;
+            this._nameSampler = name;
             var uniform = Uniform.createInt1( unit, name );
             uniforms[ 'ShadowTexture' + unit ] = uniform;
 
+            // Per Class Uniform Cache
             obj.uniforms[ unit ] = new Map( uniforms );
 
             this.latestUnit = unit;
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowTexture.js]: this._latestUnit

File: sources/osgShadow/ShadowMap.js (hunk 18 above)
Comment (cedricpinson) [inline sources/osgShadow/ShadowMap.js]: I am not sure about passing function or class instance in settings, it's not used for all other osgjs code.

Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: I started with only the apply of the default visitor calling a callback and in the end wasn't flexible enough, so I put the entire thing as a visitor.

But it could be a callback and give the shadow casted scene root node as parameter, and then user decide what he does with it.

File: sources/osgShadow/ShadowTexture.js (hunk 19 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowTexture.js]: fixed

File: sources/osgShadow/ShadowMap.js (hunk 18 above)
Comment (Kuranes) [inline sources/osgShadow/ShadowMap.js]: removed for now, will fix in the appropried Branch/PR: https://github.com/cedricpinson/osgjs/tree/feature/shadowcastervisitor

File: sources/osgShader/Compiler.js [hunk 20]
```diff
@@ -775,6 +783,34 @@ define( [
...
+
+            // per texture uniforms
+            var shadowTextureUniforms = this.getOrCreateTextureStateAttributeUniforms( shadowTexture, 'shadowTexture', tUnit );
+
+            // UGLY REMOVAL
+            // TODO: create a specific texture getuniform for compiler
+            var id = shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ].getID();
+            shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ] = undefined;
+            this._variables[ shadowTexSamplerName ] = undefined;
+            this._activeNodeList[ id ] = undefined;
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: this line seems useless because just after we delete it entry

File: sources/osgShader/Compiler.js [hunk 21]
```diff
@@ -775,6 +783,34 @@ define( [
...
 
+        createShadowTextureInputVarying: function ( shadowTexture, inputs, vertexWorld, tUnit ) {
+            var shadowTexSamplerName = 'Texture' + tUnit;
+
+            // per texture uniforms
+            var shadowTextureUniforms = this.getOrCreateTextureStateAttributeUniforms( shadowTexture, 'shadowTexture', tUnit );
+
+            // UGLY REMOVAL
+            // TODO: create a specific texture getuniform for compiler
+            var id = shadowTextureUniforms[ 'shadowTexture' + shadowTexSamplerName ].getID();
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: is that part need an update because of sampler named changed ?

Comment (Kuranes) [inline sources/osgShader/Compiler.js]: no need, didn't change other uniform name, juste texture name ( it's the shadowTexSamplerName 2 line above I had to change)

File: sources/osgShadow/ShadowTexture.js [hunk 22]
```diff
@@ -77,17 +77,24 @@ define( [
...
+            uniforms[ name ] = uniform;
 
+            // Per Class Uniform Cache
             obj.uniforms[ unit ] = new Map( uniforms );
 
-            this.latestUnit = unit;
             return obj.uniforms[ unit ];
         },
 
+        getSamplerName: function () {
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowTexture.js]: I guess we can remove it ?

Comment (Kuranes) [inline sources/osgShadow/ShadowTexture.js]: yes

File: sources/osgShader/Compiler.js (hunk 21 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: added a explanation comment

File: sources/osgShader/Compiler.js (hunk 20 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: removed
